packages/headers-parser/src/merge.ts

```diff
@@ -19,14 +17,33 @@ import { splitResults } from './results.js'
 export const mergeHeaders = function ({ fileHeaders, configHeaders }) {
   const results = [...fileHeaders, ...configHeaders]
   const { headers, errors } = splitResults(results)
-  const mergedHeaders = headers.filter(isUniqueHeader)
+  const mergedHeaders = removeDuplicates(headers)
   return { headers: mergedHeaders, errors }
 }
 
 // Remove duplicates. This is especially likely considering `fileHeaders` might
 // have been previously merged to `configHeaders`, which happens when
 // `netlifyConfig.headers` is modified by plugins.
-// The latest duplicate value is the one kept.
-const isUniqueHeader = function (header, index, headers) {
-  return !headers.slice(index + 1).some((otherHeader) => isDeepStrictEqual(header, otherHeader))
+// The latest duplicate value is the one kept hence why we reverse the arrays at
+// the beginning and at the end
+const removeDuplicates = function (headers) {
+  const uniqueHeaders = new Set()
+  const result = []
+  headers.reverse().forEach((h) => {
+    const key = generateHeaderKey(h)
+    if (uniqueHeaders.has(key)) return
+    uniqueHeaders.add(key)
+    result.push(h)
+  })
+  return result.reverse()
```

Comment:
```suggestion
  const uniqueHeaders = new Map<string, unknown>();
  for (let i = redirects.length - 1; i >= 0; i--) {
    const key = stringify.default.stableStringify(redirects[i])
    if (!uniqueHeaders.has(key)) {
      uniqueHeaders.set(key, redirects[i])
    }
  })
  return [...uniqueRedirects.values()].reverse()
```


```suggestion
  for (let i = redirects.length - 1; i >= 0; i--) {
    const r = redirects[i]
    const key = stringify.default.stableStringify(r)
    if (uniqueRedirects.has(key)) {
      uniqueRedirects.add(key)
      result.
    }
    if (uniqueHeaders.has(key)) return
    uniqueHeaders.add(key)
    result.push(h)
  })
  return result.reverse()
}
```

Reply:
Re-using the same process as in https://github.com/netlify/build/pull/4652